editor/components/post-publish-panel/toggle.js

```diff
@@ -2,28 +2,41 @@
  * External Dependencies
  */
 import { connect } from 'react-redux';
+import { get } from 'lodash';
 
 /**
  * WordPress Dependencies
  */
-import { Button, Dashicon } from '@wordpress/components';
+import { Button, withAPIData } from '@wordpress/components';
+import { compose } from '@wordpress/element';
 
 /**
  * Internal Dependencies
  */
 import PostPublishButtonLabel from '../post-publish-button/label';
+import PostPublishButton from '../post-publish-button';
 import {
 	isSavingPost,
 	isEditedPostSaveable,
 	isEditedPostPublishable,
 	isCurrentPostPublished,
+	isEditedPostBeingScheduled,
+	getCurrentPostType,
 } from '../../store/selectors';
 
-function PostPublishPanelToggle( { isSaving, isPublishable, isSaveable, isPublished, onToggle, isOpen } ) {
+function PostPublishPanelToggle( { user, isSaving, isPublishable, isSaveable, isPublished, isBeingScheduled, onToggle, isOpen } ) {
 	const isButtonEnabled = (
 		! isSaving && isPublishable && isSaveable
 	) || isPublished;
 
+	const userCanPublishPosts = get( user.data, [ 'post_type_capabilities', 'publish_posts' ], false );
+	const isContributor = user.data && ! userCanPublishPosts;
+	const showToggle = ! isContributor && ! isPublished && ! isBeingScheduled;
+
+	if ( ! showToggle ) {
```

Comment:
It feels a little weird to me that rendering one component (`PostPublishPanelToggle`) might render a completely different component (`PostPublishButton`).

Perhaps this logic should be hoisted out into `Header` or into a new wrapper component?

Reply:
The idea is to avoid using the `post` selectors in the `edit-post` directory which should be concerned only about UI and should be moved to a separate module in the future.

But you're right UI and state are a bit mixed here but that's the "balance" I've found for now.
Maybe I should rename the component something like `PublishToggleOrUpdate` :)

Reply:
Once the `data` module is updated with state exposing capabilities, we could move some of this logic outside the components folder and still access exposed data from the `editor` module.